# [1] Import Library

In [1]:
import pandas as pd
import os

# [2] Data Wrangling

## [2.1] Gathering Data

In [29]:
df = pd.read_csv('./Raw Data/product/product_dataset.csv')

In [30]:
df.head()

jenis        kategori jenis_kulit             merk  \
0  Cleanser           Toner      Kering  Viva Cosmetics    
1  Cleanser           Toner      Kering         Whitelab   
2  Cleanser           Toner      Kering            NPURE   
3  Cleanser  Makeup Remover      Kering        Somethinc   
4  Cleanser  Makeup Remover      Kering            Emina   

                            nama_product  \
0                              Air Mawar   
1                      Brightening Toner   
2         Marigold Clearings Petal Toner   
3  Reset Gentle Micellar Cleansing Water   
4              Skin Buddy Micellar Water   

                                         link_gambar  \
0  https://image.femaledaily.com/dyn/210/images/p...   
1  https://image.femaledaily.com/dyn/210/images/p...   
2  https://image.femaledaily.com/dyn/210/images/p...   
3  https://image.femaledaily.com/dyn/210/images/p...   
4  https://image.femaledaily.com/dyn/210/images/p...   

                                   nama_gambar  \
0                       vivacosmetics_airmawar   
1                    whitelab_brighteningtoner   
2            npure_marigoldclearingspetaltoner   
3  somethinc_resetgentlemicellarcleansingwater   
4                 emina_skinbuddymicellarwater   

                                     key_ingredients  \
0                                                NaN   
1                                                NaN   
2  Anti-acne: Niacinamide\nCell-communicating ing...   
3  Anti-acne: Zinc Gluconate\nAntioxidant: Calend...   
4  Antioxidant: Calendula Officinalis Flower Extr...   

                                         ingredients  \
0  Aqua, Ethanol, Methylparaben, Rose Extract, Pr...   
1  Aqua, Ethyl Alcohol, Propylene Glycol, Glyceri...   
2  Aqua, Glycerin, Lactic Acid, Tromethamine, Pro...   
3  Water, PEG-6 Caprylic/​Capric Glycerides, Prop...   
4  Aqua, Propanediol, PEG-6 Caprylic/​Capric Glyc...   

                                           deskripsi        no_BPOM  \
0  Air Mawar mengandung sari bunga mawar dan arom...  NA18121202046   
1  Whitelab Brightening Toner adalah toner yang b...  NA18221205551   
2  Toner dengan kelopak bunga Marigold asli yang ...  NA18221202239   
3  Vegan micellar water dengan surfaktan non-ioni...  NA18211208373   
4  Micellar water yang dapat mengangkat make up d...  NA18201203986   

  hamil_menyusui  kondisi_kulit sensitif tujuan  \
0             Ya            NaN        b      a   
1             Ya            NaN        b      a   
2          Tidak            NaN        b      a   
3          Tidak            NaN        b      a   
4             Ya            NaN        b      a   

                                            kegunaan fungsi  
0  Melembabkan kulit, Menyegarkan kulit, Anti-inf...    a,b  
1  Mencerahkan kulit, Mengangkat sisa kotoran di ...    d,e  
2  Menutrisi kulit, Mencegah penuaan dini, Melind...  a,c,d  
3  Membersihkan kotoran dan minyak wajah berlebih...      g  
4  Mengangkat make up dan kotoran, menyegarkan kulit      e

Insight:
1. Ada 17 kolom berbeda
2. Nilai untuk tiap kolom adalah sebagai berikut:
    - jenis = jenis skincare
    - kategori = kategori skincare
    - jenis_kulit = jenis kulit
    - merk = merk skincare
    - nama_product = nama produk
    - link_gambar = sumber gambar produk yang digunakan
    - nama_gambar = nama dari gambar yang telah diunduh
    - key_ingredients = ingredients utama dari skincare
    - ingredients = ingredient atau bahan penyusun skincare
    - deskripsi = deskripsi produk skincare terkait
    - no_BPOM = nomor bpom dari skincare
    - hamil_menyusui = kolom penanda apakah skincare terkait boleh atau tidak digunakan oleh ibu hamil atau menyusui
    - kondisi_kulit = kosong, tidak digunakan
    - sensitif = kolom yang menandakan apakah skincare dapat digunakan oleh pemilik kulit sensitif
        - a = sensitif
        - b = tidak sensitif
    - tujuan = kolom yang menandakan label tujuan dari skincare, tujuan bisa dibedakan menjadi berikut:
        - a. Membersihkan wajah dari kotoran, minyak, sisa makeup, dan debu
        - b. Mengatasi masalah kulit seperti jerawat, noda hitam, kulit kusam, atau tanda penuaan
        - c. Menghidrasi kulit, pembersihan mendalam, dan menenangkan kulit
        - d. Menjaga kelembapan kulit, mencegah kulit kering, dan mengurangi resiko iritasi tau kerusakan kulit
    - kegunaan = kegunaan dari produk terkait (dalam bentuk kalimat atau teks pendek) yang umumnya merupakan kegunaan pertama yang tergantung pada jenis skincare. misalnya cleanser, maka akan cenderung memiliki kegunaan untuk membersihkan
    - fungsi = kegunaan tambahan di luar kegunaan utama dari produk terkait. digunakan pada bagian asesmen, fungsi dibedakan menjadi berikut:
        - a. Anti-aging
        - b. Melembapkan kulit
        - c. Menutrisi kulit
        - d. Mencerahkan kulit
        - e. Menyegarkan kulit
        - f. Menenangkan kulit
        - g. Tidak, hanya fungsi Utama

Note: 
1. kolom nomor 12, 14, 15, dan 17 (indeks 11, 13, 14, dan 16) digunakan untuk asesmen sehingga dapat dipertimbangkan untuk dihapus agar tidak ditambahkan pada database
2. kolom nomor 13 dan 6 (indeks 12 dan 5) tidak digunakan, drop
3. kolom nomor 6, 7, 8, 10, 11, dan 16 (indek 5, 6, 7, 9, 10, dan 15) tidak digunakan untuk asesmen sehingga dapat dipertimbangkan untuk dihapus

In [31]:
folder = './Photos'
image_extensions = ['.jpg', '.png', '.jpeg', '.bmp', '.gif']
image_files = [f for f in os.listdir(folder) if os.path.splitext(f)[1].lower() in image_extensions]
image_files_no_ext = [os.path.splitext(f)[0] for f in image_files]
print("Daftar nama gambar:")
print(image_files_no_ext)
print("Banyaknya gambar:")
print(len(image_files_no_ext))

Daftar nama gambar:
['acnes_creamywash', 'acnes_sealinggel', 'anua_heartleafporecontrolcleansingoil', 'aveeno_dailymoisturizinglotion', 'aveeno_skinreliefmoisturizinglotion', 'avene_thermalspringwater', 'avoskin_hydratingtreatmentessence', 'avoskin_miraculousacnesolutionspottreatment', 'avoskin_miraculousdivineauratoner-essence', 'avoskin_miraculousrefiningserum', 'avoskin_perfecthydratingtreatmentessence', 'avoskin_serummiraculousrefining', "avoskin_yourskinbaeniacinami'de", 'axis-y_mugwortporeclarifyingwashoffpack', 'azarinecosmetic_purifyingdeepcleansingclaymask', 'azarine_calmmyacnesunscreenmoisturiserspf35pa+++', 'azarine_cwhitenightmoisturizer', 'azarine_hydrasoothesunscreengelspf45+++', 'azarine_hydrasoothesunscreengelspf45pa++++', 'beautyofjoseon_centellaasiaticacalmingmask', 'beautyofjoseon_ginsengcleansingoil', 'benton_fermentationeyecream', 'benton_honesteyelashserum', 'bioaqua_redbeanmoisturizing&brighteningfacialmask', 'bioaqua_salicylicacidacne', 'biobeautylab_luxuriousfa

Insight:
1. Terdapat 223 gambar

## [2.2] Assessing Data

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 302 entries, 0 to 301
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   jenis            302 non-null    object 
 1   kategori         302 non-null    object 
 2   jenis_kulit      302 non-null    object 
 3   merk             287 non-null    object 
 4   nama_product     287 non-null    object 
 5   link_gambar      287 non-null    object 
 6   nama_gambar      302 non-null    object 
 7   key_ingredients  284 non-null    object 
 8   ingredients      287 non-null    object 
 9   deskripsi        287 non-null    object 
 10  no_BPOM          232 non-null    object 
 11  hamil_menyusui   302 non-null    object 
 12  kondisi_kulit    0 non-null      float64
 13  sensitif         302 non-null    object 
 14  tujuan           302 non-null    object 
 15  kegunaan         287 non-null    object 
 16  fungsi           302 non-null    object 
dtypes: float64(1), o

Insight:
1. Kolom kondisi kulit null semua
2. Kolom merk, nama_product, ingredients, deskripsi, dan kegunaan memiliki jumlah data non null yang sama
    Note:
    Dapat diasumsikan ada 287 produk, di mana 302 - 287 = 15 data kosong jika dibandingkan kategori dan jenis kulit. Artinya, ada kategori dari jenis kulit tertentu atau sebaliknya yang tidak ditemukan produk yang cocok untuk direkomendasikan
3. Kolom no_BPOM harus berjumlah = poin nomor 2 untuk membuktikan bahwa produk layak direkomendasikan dan digunakan
4. Kolom hamil_menyusui tidak memiliki jumlah yang sama dengan poin 2, artinya ada produk yang tidak ditemukan informasi mengenai kelayakan digunakan oleh ibu hamil atau menyusui
5. Sesuai dengan poin 1, ada 287 product namun gambar yang ada hanyalah 223. Pastikan semua nama gambar ada di dalam folder photos

Note:
1. Untuk memastikan hanya produk yang layak atau teruji klinis yang direkomendasikan, maka drop semua produk yang tidak ditemukan nomor BPOMnya
2. Karena data tidak digunakan untuk analisis data lebih lanjut (misalnya visuallisasi ataupun groupby), maka tipe data tidak diubah. Jika diubah, maka beberapa kolom yang memiliki nilai berpola atau terbatas bisa diubah menjadi kategori
3. Pastikan semua nama gambar ada di dalam folder photos

In [33]:
df.isna().sum()

jenis                0
kategori             0
jenis_kulit          0
merk                15
nama_product        15
link_gambar         15
nama_gambar          0
key_ingredients     18
ingredients         15
deskripsi           15
no_BPOM             70
hamil_menyusui       0
kondisi_kulit      302
sensitif             0
tujuan               0
kegunaan            15
fungsi               0
dtype: int64

Insight:
1. Seperti yang telah disebutkan sebelumnya, ada 15 merk, nama_product, link_gambar yang null
2. Key ingredients memiliki jumlah null lebih banyak (18) jika dibandingkan poin nomor 1. Artinya ada 3 product tanpa key ingredient
3. no_BPOM banyak yang null, harus didrop
4. Ingredients, deskripsi, kegunaan memiliki null yang sama dengan poin 1, benar

Note:
1. Drop baris yang tidak berisi nama product
2. Drop semua baris dengan no_BPOM null

In [34]:
df.duplicated().sum()

6

Insight:
1. Terdapat 6 duplikat

Note:
1. Drop semua duplikat

In [35]:
duplicates = df[df.duplicated()]
print(duplicates)

           jenis                 kategori jenis_kulit merk nama_product  \
248     Cleanser           Cream & Lotion  Berjerawat  NaN          NaN   
262    Treatment  Skin Soothing Treatment  Berjerawat  NaN          NaN   
263    Treatment  Skin Soothing Treatment  Berjerawat  NaN          NaN   
265    Treatment    Brow & Lash Treatment  Berjerawat  NaN          NaN   
266    Treatment    Brow & Lash Treatment  Berjerawat  NaN          NaN   
289  Moisturizer                Face Mist  Berjerawat  NaN          NaN   

    link_gambar nama_gambar key_ingredients ingredients deskripsi no_BPOM  \
248         NaN           _             NaN         NaN       NaN     NaN   
262         NaN           _             NaN         NaN       NaN     NaN   
263         NaN           _             NaN         NaN       NaN     NaN   
265         NaN           _             NaN         NaN       NaN     NaN   
266         NaN           _             NaN         NaN       NaN     NaN   
289         

Insight:
1. Yang terduplikat adalah data null yang tidak akan digunakan

Note:
1. Drop semua duplikat

In [36]:
df.describe()

kondisi_kulit
count            0.0
mean             NaN
std              NaN
min              NaN
25%              NaN
50%              NaN
75%              NaN
max              NaN

Insight:
1. Tidak ada informasi yang diperoleh karena tipe datanya adalah object dan describe() digunakan untuk menampilkan statistik numerik

In [37]:
unique_values = df['hamil_menyusui'].unique()
print(unique_values)

['Ya' 'Tidak'
 'Bahan yang kami formulasikan sudah mendapat persetujuan standar BPOM. Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & perubahan hormonal, lebih baik konsultasikan kandungan bahan dalam produk kepada dokter kandungan anda.'
 'Ibu menyusui atau ibu hamil disarankan untuk berkonsultasi dulu dengan dokter jika ingin menggunakan produk ini.'
 'Harap Konsultasi terlebih dahulu' 'Harap konsultasi terlebih dahulu'
 'Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & mengalami perubahan hormonal, disarankan untuk berkonsultasi ke dokter kandungan mengenai bahan aktif dalam skincare yang digunakan.'
 'Ys']


Insight:
1. Ada banyak nilai yang belum bernilai 'Ya' atau 'Tidak'

Note:
1. Perbaiki kesalahan menjadi sebagai berikut:
    'Bahan yang kami formulasikan sudah mendapat persetujuan standar BPOM. Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & perubahan hormonal, lebih baik konsultasikan kandungan bahan dalam produk kepada dokter kandungan anda.': 'Ya',
    'Ibu menyusui atau ibu hamil disarankan untuk berkonsultasi dulu dengan dokter jika ingin menggunakan produk ini.': 'Tidak',
    'Harap Konsultasi terlebih dahulu': 'Tidak',
    'Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & mengalami perubahan hormonal, disarankan untuk berkonsultasi ke dokter kandungan mengenai bahan aktif dalam skincare yang digunakan.': 'Ya',
    'Ys': 'Ya',
    'Harap konsultasi terlebih dahulu' : 'Tidak'
2. Karena asesmen berupa pilihan ganda, maka ubah nilai 'Ya' menjadi 'a' dan 'Tidak' menjadi 'b'

In [38]:
unique_jenis = df['jenis'].unique()
print(unique_jenis)

['Cleanser' 'Treatment' 'Mask' 'Moisturizer']


Insight:
1. Ada 4 nilai jenis, benar

In [39]:
unique_kategori = df['kategori'].unique()
print(unique_kategori)


['Toner' 'Makeup Remover' 'Cream & Lotion' 'Facial Wash' 'Oil'
 'Scrub & Exfoliator' 'Eye Treatment' 'Skin Soothing Treatment'
 'Brow & Lash Treatment' 'Peeling' 'Serum & Essence' 'Acne Treatment'
 'Sleeping Mask' 'Nose Pack' 'Mask Sheet' 'Wash-Off' 'Lotion & Emulsion'
 'Face Mist' 'Gel' 'Sun Protection' 'Cream' 'Face Oil']


Insight:
1. Terdapat 22 kategori skincare, yaitu 'Toner' 'Makeup Remover', 'Cream & Lotion', 'Facial Wash', 'Oil', 'Scrub & Exfoliator', 'Eye Treatment', 'Skin Soothing Treatment', 'Brow & Lash Treatment' 'Peeling' 'Serum & Essence' 'Acne Treatment', 'Sleeping Mask', 'Nose Pack', 'Mask Sheet', 'Wash-Off', 'Lotion & Emulsion', 'Face Mist', 'Gel', 'Sun Protection', 'Cream', dan 'Face Oil'

In [40]:
unique_kulit = df['jenis_kulit'].unique()
print(unique_kulit)

['Kering' 'Berminyak' 'Sensitif' 'Normal' 'Berjerawat']


Insight:
1. Terdapat 5 kategori kulit, yaitu Kering, Berminyak, Sensitif, Normal, dan Berjerawat

In [41]:
unique_sensitif = df['sensitif'].unique()
print(unique_sensitif)

['b' 'a']


Insight:
1. Terdapat 2 tipe nilai pada kolom sensitif yaitu b yang mewakili 'Tidak sensitif' dan a yang mewakili 'Sensitif'

In [42]:
unique_tujuan = df['tujuan'].unique()
print(unique_tujuan)

['a' 'b' 'c' 'd']


Insight:
1. Terdapat 4 tipe nilai pada kolom tujuan yang mewakili:
    a. Membersihkan wajah dari kotoran, minyak, sisa makeup, dan debu
    b. Mengatasi masalah kulit seperti jerawat, noda hitam, kulit kusam, atau tanda penuaan
    c. Menghidrasi kulit, pembersihan mendalam, dan menenangkan kulit
    d. Menjaga kelembapan kulit, mencegah kulit kering, dan mengurangi resiko iritasi tau kerusakan kulit

In [43]:
unique_huruf = df['fungsi'].str.split(',').explode().unique()
print(unique_huruf)

['a' 'b' 'd' 'e' 'c' 'g' 'f']


Insight:
1. Terdapat banyak kombinasi nilai dari kolom fungsi. Walaupun hurufnya sama, pembedanya adalah ada tidaknya spasi di bagian depan

Note:
1. Perbaiki nilai, putuskan apakah berisi spasi sebelum huruf atau tidak

Kesimpulan:
1. kolom nomor 12, 14, 15, dan 17 (indeks 11, 13, 14, dan 16) digunakan untuk asesmen sehingga dapat dipertimbangkan untuk dihapus agar tidak ditambahkan pada database
2. kolom nomor 13 dan 6 (indeks 12 dan 5) tidak digunakan, drop
3. kolom nomor 6, 7, 8, 9, 10, 11, dan 16 (indek 5, 6, 7, 8, 9, 10, dan 15) tidak digunakan untuk asesmen sehingga dapat dipertimbangkan untuk dihapus
4. untuk memastikan hanya produk yang layak atau teruji klinis yang direkomendasikan, maka drop semua produk yang tidak ditemukan nomor BPOMnya
5. drop semua duplikat
6. perbaiki kesalahan menjadi sebagai berikut:
    'Bahan yang kami formulasikan sudah mendapat persetujuan standar BPOM. Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & perubahan hormonal, lebih baik konsultasikan kandungan bahan dalam produk kepada dokter kandungan anda.': 'Ya',
    'Ibu menyusui atau ibu hamil disarankan untuk berkonsultasi dulu dengan dokter jika ingin menggunakan produk ini.': 'Tidak',
    'Harap Konsultasi terlebih dahulu': 'Tidak',
    'Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & mengalami perubahan hormonal, disarankan untuk berkonsultasi ke dokter kandungan mengenai bahan aktif dalam skincare yang digunakan.': 'Ya',
    'Ys': 'Ya',
    'Harap konsultasi terlebih dahulu' : 'Tidak'
7. karena asesmen berupa pilihan ganda, maka ubah nilai 'Ya' menjadi 'a' dan 'Tidak' menjadi 'b'
8. pastikan semua nama gambar ada di dalam folder photos
9. drop baris yang tidak ada nama product
10. (PENGECEKAN MANUAL) Banyak data pada kolom tertentu tidak diakhiri dengan tanda baca titik. Misalnya pada kolom ingredients, deskripsi, dan kegunaan.
11. (PENGECEKAN MANUAL) Penggunaan huruf kapital dan tidak pada kolom kegunaan masih rancu
12. (PENGECEKAN MANUAL) Ada beberapa karakter khusus yang ditemukan, ganti

Urutan Cleaning:
2, 5, 4, 9, 6, 7, 8, 10, 1, 3

Poin nomor 11 dan 12 akan dilakukan secara manual setelah cleaning selesai


## [2.3] Cleaning Data

### 2.3.1 Drop kolom nomor 13 dan 6 (indeks 12 dan 5) yaitu kondisi kulit dan link gambar

In [44]:
df.drop(columns=['link_gambar', 'kondisi_kulit'], inplace=True)

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 302 entries, 0 to 301
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   jenis            302 non-null    object
 1   kategori         302 non-null    object
 2   jenis_kulit      302 non-null    object
 3   merk             287 non-null    object
 4   nama_product     287 non-null    object
 5   nama_gambar      302 non-null    object
 6   key_ingredients  284 non-null    object
 7   ingredients      287 non-null    object
 8   deskripsi        287 non-null    object
 9   no_BPOM          232 non-null    object
 10  hamil_menyusui   302 non-null    object
 11  sensitif         302 non-null    object
 12  tujuan           302 non-null    object
 13  kegunaan         287 non-null    object
 14  fungsi           302 non-null    object
dtypes: object(15)
memory usage: 35.5+ KB


Insight:
1. Kolom link_gambar dan kondisi kulit berhasil terhapus

### 2.3.2 Drop semua duplikat

In [46]:
df = df.drop_duplicates()

In [47]:
df.duplicated().sum()

0

Insight:
1. Sudah tidak ada yang duplikat

### 2.3.3 Drop semua yang tidak ada no_BPOM

In [48]:
df = df.dropna(subset=['no_BPOM'])

In [49]:
null_bpom = df['no_BPOM'].isnull().sum()
print(null_bpom)

0


Insight:
1. Sudah tidak ada baris dengan no_BPOM null

### 2.3.4 Drop baris yang tidak ada nama produknya

In [50]:
df = df.dropna(subset=['nama_product'])

In [51]:
null_nama_product = df['nama_product'].isnull().sum()
print(null_nama_product)

0


Insight:
1. Sudah tidak ada baris dengan nama_product null

### 2.3.5 Perbaiki kesalahan pada kolom hamil_menyusui

In [52]:
replacement_dict = {
    'Bahan yang kami formulasikan sudah mendapat persetujuan standar BPOM. Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & perubahan hormonal, lebih baik konsultasikan kandungan bahan dalam produk kepada dokter kandungan anda.': 'Ya',
    'Ibu menyusui atau ibu hamil disarankan untuk berkonsultasi dulu dengan dokter jika ingin menggunakan produk ini.': 'Tidak',
    'Harap Konsultasi terlebih dahulu': 'Tidak',
    'Namun mengingat ibu hamil terkadang memiliki alergi terhadap hal-hal tertentu & mengalami perubahan hormonal, disarankan untuk berkonsultasi ke dokter kandungan mengenai bahan aktif dalam skincare yang digunakan.': 'Ya',
    'Ys': 'Ya',
    'Harap konsultasi terlebih dahulu' : 'Tidak'
}

df['hamil_menyusui'] = df['hamil_menyusui'].replace(replacement_dict)

In [53]:
unique_values = df['hamil_menyusui'].unique()
print(unique_values)

['Ya' 'Tidak']


Insight:
1. Sudah hanya bernilai 'Ya' dan 'Tidak'

### 2.3.6 Ubah 'Ya' menjadi 'a' dan 'Tidak' menjadi 'b' pada kolom hamil_menyusui

In [54]:
replace = {
    'Ya' : 'a',
    'Tidak': 'b'
}

df['hamil_menyusui'] = df['hamil_menyusui'].replace(replace)

In [55]:
unique_values = df['hamil_menyusui'].unique()
print(unique_values)

['a' 'b']


Insight:
1. Sudah hanya bernilai 'a' dan 'b'

### 2.3.7 Pastikan semua nama_gambar ada di folder Photos

In [56]:
missing_in_image_files = set(df['nama_gambar']) - set(image_files_no_ext)

print("Nama gambar yang tidak ada di folder Photos:")
print(missing_in_image_files)
print(len(missing_in_image_files))

Nama gambar yang tidak ada di folder Photos:
{'skintific_5xceramidelowphcleanser\n', 'glad2glow_centellaallantoinsoothingmoisturizergel\n', 'skin1004_madagascarcentellabhafoam\n', 'azarine_hydrasoothesunscreengelspf45pa++++\n'}
4


Insight:
1. Terdapat kesalahan yaitu pada akhir \n sehingga nama gambar tidak ditemukan

Note:
1. Telah dicari manual dan nama gambar ditemukan

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 232 entries, 0 to 301
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   jenis            232 non-null    object
 1   kategori         232 non-null    object
 2   jenis_kulit      232 non-null    object
 3   merk             232 non-null    object
 4   nama_product     232 non-null    object
 5   nama_gambar      232 non-null    object
 6   key_ingredients  229 non-null    object
 7   ingredients      232 non-null    object
 8   deskripsi        232 non-null    object
 9   no_BPOM          232 non-null    object
 10  hamil_menyusui   232 non-null    object
 11  sensitif         232 non-null    object
 12  tujuan           232 non-null    object
 13  kegunaan         232 non-null    object
 14  fungsi           232 non-null    object
dtypes: object(15)
memory usage: 29.0+ KB


### 2.3.8 Menambahkan ekstensi file

In [58]:
def add_extension(nama_gambar):
    if nama_gambar in image_files_no_ext:
        index = image_files_no_ext.index(nama_gambar)
        return image_files[index]
    else:
        return None

df['nama_gambar'] = df['nama_gambar'].apply(add_extension)

print(df['nama_gambar'])


0                             vivacosmetics_airmawar.jpg
1                          whitelab_brighteningtoner.jpg
2                  npure_marigoldclearingspetaltoner.jpg
3        somethinc_resetgentlemicellarcleansingwater.jpg
4                       emina_skinbuddymicellarwater.jpg
                             ...                        
297                   whitelab_brighteningnightcream.jpg
298          somethinc_calmdown!skinpairr-covercream.jpg
299    theoriginote_acneb5serumsalicylicacid+pantheno...
300                naturmiracle_antiacnefaceoilserum.jpg
301             biobeautylab_luxuriousfacialoilserum.jpg
Name: nama_gambar, Length: 232, dtype: object


Insight:
1. Nama gambar telah ditambahkan ekstensi aslinya

In [59]:
df['nama_gambar'] = df['nama_gambar'].astype(str)

def check_no_extension(file_name):
    return '.' not in file_name

df['tanpa_ekstensi'] = df['nama_gambar'].apply(check_no_extension)
gambar_tanpa_ekstensi = df[df['tanpa_ekstensi']]
jumlah_tanpa_ekstensi = len(gambar_tanpa_ekstensi)
print(f"Jumlah gambar tanpa ekstensi: {jumlah_tanpa_ekstensi}")
print("Nama gambar tanpa ekstensi:")
print(gambar_tanpa_ekstensi['nama_gambar'].tolist())


Jumlah gambar tanpa ekstensi: 4
Nama gambar tanpa ekstensi:
['None', 'None', 'None', 'None']


Insight:
1. Ada 4 gambar tanpa ekstensi, namun ketika dicek tertulis 'none' sehingga akan dilakukan cek dan perbaikan manual

### 2.3.9 Menambahkan titik di akhir kalimat

In [60]:
def add_period_if_missing(text):
    if isinstance(text, str):
        return text if text.endswith('.') else text + '.'
    return text


In [61]:
df['ingredients'] = df['ingredients'].apply(add_period_if_missing)

In [62]:
df['deskripsi'] = df['deskripsi'].apply(add_period_if_missing)

In [63]:
df['kegunaan'] = df['kegunaan'].apply(add_period_if_missing)

### 2.3.10 Save CSV untuk Database

In [69]:
df_database = df.copy()

In [70]:
df_database.head()

jenis        kategori jenis_kulit             merk  \
0  Cleanser           Toner      Kering  Viva Cosmetics    
1  Cleanser           Toner      Kering         Whitelab   
2  Cleanser           Toner      Kering            NPURE   
3  Cleanser  Makeup Remover      Kering        Somethinc   
4  Cleanser  Makeup Remover      Kering            Emina   

                            nama_product  \
0                              Air Mawar   
1                      Brightening Toner   
2         Marigold Clearings Petal Toner   
3  Reset Gentle Micellar Cleansing Water   
4              Skin Buddy Micellar Water   

                                       nama_gambar  \
0                       vivacosmetics_airmawar.jpg   
1                    whitelab_brighteningtoner.jpg   
2            npure_marigoldclearingspetaltoner.jpg   
3  somethinc_resetgentlemicellarcleansingwater.jpg   
4                 emina_skinbuddymicellarwater.jpg   

                                     key_ingredients  \
0                                                NaN   
1                                                NaN   
2  Anti-acne: Niacinamide\nCell-communicating ing...   
3  Anti-acne: Zinc Gluconate\nAntioxidant: Calend...   
4  Antioxidant: Calendula Officinalis Flower Extr...   

                                         ingredients  \
0  Aqua, Ethanol, Methylparaben, Rose Extract, Pr...   
1  Aqua, Ethyl Alcohol, Propylene Glycol, Glyceri...   
2  Aqua, Glycerin, Lactic Acid, Tromethamine, Pro...   
3  Water, PEG-6 Caprylic/​Capric Glycerides, Prop...   
4  Aqua, Propanediol, PEG-6 Caprylic/​Capric Glyc...   

                                           deskripsi        no_BPOM  \
0  Air Mawar mengandung sari bunga mawar dan arom...  NA18121202046   
1  Whitelab Brightening Toner adalah toner yang b...  NA18221205551   
2  Toner dengan kelopak bunga Marigold asli yang ...  NA18221202239   
3  Vegan micellar water dengan surfaktan non-ioni...  NA18211208373   
4  Micellar water yang dapat mengangkat make up d...  NA18201203986   

  hamil_menyusui sensitif tujuan  \
0              a        b      a   
1              a        b      a   
2              b        b      a   
3              b        b      a   
4              a        b      a   

                                            kegunaan fungsi  tanpa_ekstensi  
0  Melembabkan kulit, Menyegarkan kulit, Anti-inf...    a,b           False  
1  Mencerahkan kulit, Mengangkat sisa kotoran di ...    d,e           False  
2  Menutrisi kulit, Mencegah penuaan dini, Melind...  a,c,d           False  
3  Membersihkan kotoran dan minyak wajah berlebih...      g           False  
4  Mengangkat make up dan kotoran, menyegarkan ku...      e           False

In [71]:
df_database = df_database.drop(columns=['hamil_menyusui', 'sensitif', 'tujuan', 'fungsi', 'tanpa_ekstensi'])
df_database.info()


<class 'pandas.core.frame.DataFrame'>
Index: 232 entries, 0 to 301
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   jenis            232 non-null    object
 1   kategori         232 non-null    object
 2   jenis_kulit      232 non-null    object
 3   merk             232 non-null    object
 4   nama_product     232 non-null    object
 5   nama_gambar      232 non-null    object
 6   key_ingredients  229 non-null    object
 7   ingredients      232 non-null    object
 8   deskripsi        232 non-null    object
 9   no_BPOM          232 non-null    object
 10  kegunaan         232 non-null    object
dtypes: object(11)
memory usage: 21.8+ KB


Insight:
1. Dataframe baru bernama df_database sudah hanya berisi kolom yang akan disimpan pada database
2. Bagian key_ingredients ada 3 yang null

Note:
1. Tidak perlu mendrop key_ingredients atau menambahkan informasi untuk key_ingredients karena informasi key_ingredients yang memang terbatas dan ingredients telah disampaikan dengan lengkap pada kolom ingredients

In [74]:
df_database.to_csv('./Clean Data/product_database.csv', index=False)

### 2.3.11 Save CSV untuk Asesmen

In [75]:
df_asesmen = df.copy()

In [76]:
df_asesmen.head()

jenis        kategori jenis_kulit             merk  \
0  Cleanser           Toner      Kering  Viva Cosmetics    
1  Cleanser           Toner      Kering         Whitelab   
2  Cleanser           Toner      Kering            NPURE   
3  Cleanser  Makeup Remover      Kering        Somethinc   
4  Cleanser  Makeup Remover      Kering            Emina   

                            nama_product  \
0                              Air Mawar   
1                      Brightening Toner   
2         Marigold Clearings Petal Toner   
3  Reset Gentle Micellar Cleansing Water   
4              Skin Buddy Micellar Water   

                                       nama_gambar  \
0                       vivacosmetics_airmawar.jpg   
1                    whitelab_brighteningtoner.jpg   
2            npure_marigoldclearingspetaltoner.jpg   
3  somethinc_resetgentlemicellarcleansingwater.jpg   
4                 emina_skinbuddymicellarwater.jpg   

                                     key_ingredients  \
0                                                NaN   
1                                                NaN   
2  Anti-acne: Niacinamide\nCell-communicating ing...   
3  Anti-acne: Zinc Gluconate\nAntioxidant: Calend...   
4  Antioxidant: Calendula Officinalis Flower Extr...   

                                         ingredients  \
0  Aqua, Ethanol, Methylparaben, Rose Extract, Pr...   
1  Aqua, Ethyl Alcohol, Propylene Glycol, Glyceri...   
2  Aqua, Glycerin, Lactic Acid, Tromethamine, Pro...   
3  Water, PEG-6 Caprylic/​Capric Glycerides, Prop...   
4  Aqua, Propanediol, PEG-6 Caprylic/​Capric Glyc...   

                                           deskripsi        no_BPOM  \
0  Air Mawar mengandung sari bunga mawar dan arom...  NA18121202046   
1  Whitelab Brightening Toner adalah toner yang b...  NA18221205551   
2  Toner dengan kelopak bunga Marigold asli yang ...  NA18221202239   
3  Vegan micellar water dengan surfaktan non-ioni...  NA18211208373   
4  Micellar water yang dapat mengangkat make up d...  NA18201203986   

  hamil_menyusui sensitif tujuan  \
0              a        b      a   
1              a        b      a   
2              b        b      a   
3              b        b      a   
4              a        b      a   

                                            kegunaan fungsi  tanpa_ekstensi  
0  Melembabkan kulit, Menyegarkan kulit, Anti-inf...    a,b           False  
1  Mencerahkan kulit, Mengangkat sisa kotoran di ...    d,e           False  
2  Menutrisi kulit, Mencegah penuaan dini, Melind...  a,c,d           False  
3  Membersihkan kotoran dan minyak wajah berlebih...      g           False  
4  Mengangkat make up dan kotoran, menyegarkan ku...      e           False

In [79]:
df_asesmen = df_asesmen.drop(columns=['nama_gambar', 'key_ingredients', 'deskripsi', 'no_BPOM', 'kegunaan', 'tanpa_ekstensi'])
df_asesmen.info()

<class 'pandas.core.frame.DataFrame'>
Index: 232 entries, 0 to 301
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   jenis           232 non-null    object
 1   kategori        232 non-null    object
 2   jenis_kulit     232 non-null    object
 3   merk            232 non-null    object
 4   nama_product    232 non-null    object
 5   ingredients     232 non-null    object
 6   hamil_menyusui  232 non-null    object
 7   sensitif        232 non-null    object
 8   tujuan          232 non-null    object
 9   fungsi          232 non-null    object
dtypes: object(10)
memory usage: 19.9+ KB


Insight:
1. Dataframe baru bernama df_asesmen sudah hanya berisi kolom yang akan digunakan untuk asesmen
2. Tidak ada yang null, semua berjumlah sama

In [81]:
df_asesmen.to_csv('./Clean Data/product_asesmen.csv', index=False)